dimod/discrete/discrete_quadratic_model.py

```diff
@@ -0,0 +1,367 @@
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import collections.abc as abc
+
+from operator import eq
+
+import numpy as np
+
+from dimod.discrete.cydiscrete_quadratic_model import cyDiscreteQuadraticModel
+
+
+__all__ = ['DiscreteQuadraticModel', 'DQM']
+
+
+# this is the third(!) variables implementation in dimod. It differs from
+# dimod.variables in that it stores its labels sparsely. It has the same
+# behaviour as the cyBQM ones, except that it rolls the logic up into a
+# object. These need to be unified.
+class _Variables(abc.Sequence, abc.Set):
+    def __init__(self):
+        self._label_to_idx = dict()
+        self._idx_to_label = dict()
+        self.stop = 0
+
+    def __contains__(self, v):
+        return v in self._label_to_idx or (isinstance(v, int)
+                                           and 0 <= v < self.stop
+                                           and v not in self._idx_to_label)
+
+    def __eq__(self, other):
+        if isinstance(other, abc.Sequence):
+            return len(self) == len(other) and all(map(eq, self, other))
+        elif isinstance(other, abc.Set):
+            return not (self ^ other)
+        else:
+            return False
+
+    def __getitem__(self, idx):
+
+        if not isinstance(idx, int):
+            raise TypeError("index must be an integer.")
+
+        given = idx  # for error message
+
+        # handle negative indexing
+        if idx < 0:
+            idx = self.stop + idx
+
+        if idx >= self.stop:
+            raise IndexError('index {} out of range'.format(given))
+
+        return self._idx_to_label.get(idx, idx)
+
+    def __len__(self):
+        return self.stop
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    @property
+    def is_range(self):
+        return not self._label_to_idx
+
+    def _append(self, v=None):
+        """Append a new variable."""
+
+        if v is None:
+            # handle the easy case
+            if self.is_range:
+                self.stop += 1
+                return
+
+            raise NotImplementedError
+
+        elif v in self:
+            raise ValueError('{!r} is already a variable'.format(v))
+
+        idx = self.stop
+
+        if idx != v:
+            self._label_to_idx[v] = idx
+            self._idx_to_label[idx] = v
+
+        self.stop += 1
+
+        return
+
+    def index(self, v):
+        # todo: support start and end like list.index
+        if v not in self:
+            raise ValueError('unknown variable {!r}'.format(v))
+        return self._label_to_idx.get(v, v)
+
+
+class DiscreteQuadraticModel:
+    """A discrete quadratic model is a polynomial over discrete variables with
+    terms all of degree two or less.
+
+    Examples:
+
+        This example constructs a map coloring with Canadian provinces. To
+        solve the problem we penalize adjacent provinces having the same color.
```

Comment:
Maybe instead of glossing an explanation, provide a link to https://docs.ocean.dwavesys.com/en/stable/examples/map_coloring.html?

Reply:
I am a bit worried that it might be confusing to link the two... Maybe a wikipedia link? We'll definitely want a map coloring example using DQM as well, possibly in the same repo